giraph-core/src/main/java/org/apache/giraph/ooc/policy/MemoryEstimatorOracle.java

```diff
@@ -0,0 +1,851 @@
+      double upperBound, double[][] xValues, double[] yValues)
+      throws Exception {
+
+      boolean result = false;
+      if (coefficient[coefIndex] < lowerBound ||
+        coefficient[coefIndex] > upperBound) {
+
+        double value;
+        if (coefficient[coefIndex] < lowerBound) {
+          value = lowerBound;
+        } else {
+          value = upperBound;
+        }
+        int ptr = -1;
+        if (validColumnIndices.size() >= 1 &&
+          validColumnIndices.get(validColumnIndices.size() - 1) == coefIndex) {
+          ptr = validColumnIndices.size() - 1;
+        } else if (validColumnIndices.size() >= 2 &&
+          validColumnIndices.get(validColumnIndices.size() - 2) == coefIndex) {
+          ptr = validColumnIndices.size() - 2;
+        }
+        if (ptr != -1) {
+          if (LOG.isDebugEnabled()) {
+            LOG.debug("addRecord: coefficient at index " + coefIndex +
+              " is wrong in the regression, setting it to " + value);
+          }
+          // remove from valid column
+          validColumnIndices.remove(ptr);
+          // re-create the X matrix
+          fillXMatrix(dataSamples, validColumnIndices, xValues);
+          // adjust Y values
+          for (int i = 0; i < memorySamples.size(); ++i) {
+            yValues[i] -= value * dataSamples.get(i)[coefIndex];
+          }
+          // save new coefficient value in intermediate array
+          extreme[coefIndex] = value;
+          // re-run regression
+          mlr.newSampleData(yValues, xValues);
+          calculateRegression(coefficient, validColumnIndices, mlr);
+          result = true;
+        } else {
+          if (LOG.isDebugEnabled()) {
+            LOG.debug(
+              "addRecord: coefficient was not in the regression, " +
+                "setting it to the lower bound");
+          }
+          result = false;
+        }
+        coefficient[coefIndex] = value;
+      }
+      return result;
+    }
+
+    /**
+     * Calculates the regression.
+     * @param coefficient Array of coefficients
+     * @param validColumnIndices List of valid columns
+     * @param mlr {@link OLSMultipleLinearRegression} instance.
+     * @throws Exception
+     */
+    private static void calculateRegression(double[] coefficient,
+      Vector<Integer> validColumnIndices, OLSMultipleLinearRegression mlr)
+      throws Exception {
+
+      if (coefficient.length != validColumnIndices.size()) {
+        throw new Exception("There are " + coefficient.length +
+          " coefficients, but " + validColumnIndices.size() +
+          " valid columns in the regression");
+      }
+
+      double[] beta = mlr.estimateRegressionParameters();
+      for (int i = 0; i < coefficient.length; ++i) {
+        coefficient[i] = 0;
+      }
+      for (int i = 0; i < validColumnIndices.size(); ++i) {
+        coefficient[validColumnIndices.get(i)] = beta[i];
+      }
+      coefficient[5] = beta[validColumnIndices.size()];
```

Comment:
There is no explicit notion of "bias". "Bias" is what the user would add if he/she wants to. The returned parameters for your case should be an array of size 1. In other words, the setting you mentioned is equivalent to find "one" coefficient b where:
`[1 2 3] = b * [0 0.1 0.2]`
And the objective is to find the b where the sum of square root of errors is minimum.

When I run the exact same code as you mentioned I get an array of size one with the value "16.0" in it. Which makes sense.

The case you have had in mind can be expressed by the `x` array as follows:
`double[][] x = new double[][] {{0, 1}, {0.1, 1}, {0.2, 1}}`
or:
`double[][] x = new double[][] {{1, 0}, {1, 0.1}, {1, 0.2}}`
In the first case I get [10.000000000000002, 1.0], and in the second case I get [0.9999999999999997, 10.0]; and both make sense.

To put it more accurately, MLR can be applied in cases where you do not have any bias. If you want to have a bias, you should add "1" to all your `x` data (and the returned coefficient from MRL for this added "1" is your bias).